From c883e824f8b16a6ac953ae173cb1a22053322bc1 Mon Sep 17 00:00:00 2001
From: Lars Kellogg-Stedman <lars@oddbit.com>
Date: Fri, 18 Sep 2026 08:58:21 -0400
Subject: [PATCH 4/4] Fix segfault when attempting to insert message with no
 mailbox open

With no mailbox open, the ~U command passed this null pointer to
n_getmsglist(), which immediately executed:

    *vector = 0;

That dereferenced address zero and caused the SIGSEGV. The forwarding
code intended to allocate a local vector, but did so only after calling
n_getmsglist().
---
 src/mx/collect.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/src/mx/collect.c b/src/mx/collect.c
index 8a1981c8..76304d1e 100644
--- a/src/mx/collect.c
+++ b/src/mx/collect.c
@@ -1013,22 +1013,22 @@ a_coll_forward(char const *ms, int f){ /* {{{ */
 	int rv, *msgvec;
 	NYD_IN;
 
-	if((rv = n_getmsglist(a_coll->cc_scope, FAL0, ms, n_msgvec, 0, NIL)) < 0){
+	/* The compose path is also used without an active mailbox, in which case
+	 * the folder-owned n_msgvec has not been allocated. */
+	msgvec = su_AUTO_TALLOC(int, msgCount +1);
+	if((rv = n_getmsglist(a_coll->cc_scope, FAL0, ms, msgvec, 0, NIL)) < 0){
 		rv = n_pstate_err_no; /* XXX not really, should be handled there! */
 		goto jleave;
 	}
 	if(rv == 0){
-		*n_msgvec = first(0, MMNORM); /* TODO integrate mode into getmsglist */
-		if(*n_msgvec == 0){
+		*msgvec = first(0, MMNORM); /* TODO integrate mode into getmsglist */
+		if(*msgvec == 0){
 			n_err(_("No appropriate messages\n"));
 			rv = su_ERR_NOMSG;
 			goto jleave;
 		}
 		rv = 1;
 	}
-
-	msgvec = su_AUTO_TALLOC(int, rv +1);
-	su_mem_copy(msgvec, n_msgvec, sizeof(*msgvec) * S(uz,rv));
 	msgvec[rv] = 0;
 
 	STRUCT_ZERO(struct a_coll_quote_ctx, &cqc);
-- 
2.55.0

